### PR TITLE
XiaoW_Refactor of timeentry controllers and fix the edit history issue

### PR DESCRIPTION
**Pull request by wang9hu**
# Description
This PR along with the frontend changes aims to refactor the time entry controller for post, edit, and delete time entries. They also fix the edit time entry history not added to edit history issue.

## Related PRS (if any):
To test this backend PR you need to checkout the #[2131](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2131#issue-2234528175) frontend PR.
…

## Main changes explained:
- refactor post, edit, and delete handlers, encapsulating code blocks into functions, making the main part of the handlers clear and easy to follow
- utilize mongoose transaction to make the changes of database atomic, including timeentry and userprofile changes
- make sure edit to time could be correctly added to edit history for non-admin users

## How to test:
1. check into current branch and frontend #[2131](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2131#issue-2234528175)
2. do `npm install` and `npm run dev` to run this PR locally
3. see frontend instruction

### CODE REVIEW

**Comment by alla0810** [comment]
lots of network errors when loaded.
network error

**Comment by alla0810** [comment]
lots of network errors when loaded

image

**Comment by alla0810** [comment]
logged in as a volunteer. Works good.

logged in as a volunteer

**Review by j-corp-25 — DISMISSED** [review]
Hi @wang9hu ,

I followed instructions for  **Volunteer** and **Admin** roles. I have ran into errors for both.

When im editing on the **Admin** role. I can edit and update the time-entry but `toastify` error still shows up when I delete the entry. The entry deletes but error still pops up

Below is the error in a recorded test for Admin


https://github.com/OneCommunityGlobal/HGNRest/assets/95154804/fbd98528-26c9-4e00-b595-180d06dfeced

I tested it with Volunteer role, and I could not make the entry. I get error on the date field.
`Invalid date. Please refresh the page.`






https://github.com/OneCommunityGlobal/HGNRest/assets/95154804/029b764b-c576-4108-97ef-bef552fd740b

**Review by SRamen1999 — DISMISSED** [review]
I left my review on the frontend PR. https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2131#pullrequestreview-1995565936

**Review by aaryaneil — CHANGES_REQUESTED** [review]
Left a review on the [frontend.](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2131#pullrequestreview-1999273129)

**Review by Tiantian-C — COMMENTED** [review]
Hi,
I have tested it and leave the review at the frontend!

**Comment by ljrirene** [comment]
Please see the PR review for the front-end:https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2131#pullrequestreview-1998776357

**Review by jennZhang93 — DISMISSED** [review]
Left a review at the fronend PR: https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2131#pullrequestreview-2004901087

**Review by ptrpengdev — COMMENTED** [review]
Comment for reference. Frontend also tested.  [Frontend PR comment](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2131#issuecomment-2067036415).

File: src/controllers/timeEntryController.js [hunk 1]
```diff
@@ -103,40 +113,263 @@ const notifyTaskOvertimeEmailBody = async (
...
+ * @returns {Void}
+ */
+const updateUserprofileCategoryHrs = async (
+  fromProjectId,
+  secondsToBeRemoved,
+  toProjectId,
+  secondsToBeAdded,
+  userprofile,
+) => {
+  if (fromProjectId) {
```
[the remark this answers is not in the document]
Comment (wang9hu) [reply src/controllers/timeEntryController.js:233]: The code blocks you are referring to are not just for posting tangible or intangible time entries, but also for changing the tangibility of a time entry from tangible to intangible or vice versa. If I create a function for post new time entry, I need to create both postNewTangbleTimeEntry and postNewIntangibleTimeEntry for that, and they will also be used in the editTimeEntry controller, which kind of conflict their names, and I don't want to create another 2 same functions with different names just for editTimeEntry.

File: src/controllers/timeEntryController.js [hunk 2]
```diff
@@ -473,22 +841,27 @@ const timeEntrycontroller = function (TimeEntry) {
...
-        if (!taskId) await updateTaskIdInTimeEntry(projectId, timeEntry); // if no taskId, then it might be old time entry data that didn't separate projectId with taskId
-        const hours = Math.floor(timeEntry.totalSeconds / 3600);
-        const minutes = Math.floor((timeEntry.totalSeconds % 3600) / 60);
-        Object.assign(timeEntry, { hours, minutes, totalSeconds: undefined });
-        return timeEntry;
-      }));
+      const results = await Promise.all(
+        timeEntries.map(async (timeEntry) => {
+          timeEntry = { ...timeEntry.toObject() };
+          const { projectId, taskId } = timeEntry;
```
[the remark this answers is not in the document]
Comment (wang9hu) [reply src/controllers/timeEntryController.js:847]: Those are not within the scope of this PR, this PR focuses only on the create, edit, and delete single time entry.

[the remark this answers is not in the document]
Comment (ptrpengdev) [reply src/controllers/timeEntryController.js:847]: Got it.

**Comment by one-community** [comment]
Thank you all, merging!

**Review by one-community — APPROVED** [review]
Comments addressed

**Pull request merged**